### PR TITLE
Impose a more sensible ordering for animation graph evaluation.

### PR DESCRIPTION
**Pull request by pcwalton**
This is an updated version of #15530.

This commit changes the animation graph evaluation to be operate in a more sensible order and updates the semantics of blend nodes to conform to [the animation composition RFC]. Prior to this patch, a node graph like this:

	    ┌─────┐
	    │     │
	    │  1  │
	    │     │
	    └──┬──┘
	       │
       ┌───────┴───────┐
       │               │
       ▼               ▼
    ┌─────┐         ┌─────┐
    │     │         │     │
    │  2  │         │  3  │
    │     │         │     │
    └──┬──┘         └──┬──┘
       │               │
   ┌───┴───┐       ┌───┴───┐
   │       │       │       │
   ▼       ▼       ▼       ▼
┌─────┐ ┌─────┐ ┌─────┐ ┌─────┐
│     │ │     │ │     │ │     │
│  4  │ │  6  │ │  5  │ │  7  │
│     │ │     │ │     │ │     │
└─────┘ └─────┘ └─────┘ └─────┘

Would be evaluated as (((4 ⊕ 5) ⊕ 6) ⊕ 7), with the blend (lerp/slerp) operation notated as ⊕. As quaternion multiplication isn't commutative, this is very counterintuitive and will especially lead to trouble with the forthcoming additive blending feature (#15198).

This patch fixes the issue by changing the evaluation order to postorder, with children of a node evaluated in ascending order by node index.

To do so, this patch revamps `AnimationCurve` to be based on an *evaluation stack* and a *blend register*. During target evaluation, the graph evaluator traverses the graph in postorder. When encountering a clip node, the evaluator pushes the possibly-interpolated value onto the evaluation stack. When encountering a blend node, the evaluator pops values off the stack into the blend register, accumulating weights as appropriate. When the graph is completely evaluated, the top element on the stack is *committed* to the property of the component.

A new system, the *graph threading* system, is added in order to cache the sorted postorder traversal to avoid the overhead of sorting children at animation evaluation time. Mask evaluation has been moved to this system so that the graph only has to be traversed at most once per frame. Unlike the `ActiveAnimation` list, the *threaded graph* is cached from frame to frame and only has to be regenerated when the animation graph asset changes.

This patch currently regresses the `animate_target` performance in `many_foxes` by around 50%, resulting in an FPS loss of about 2-3 FPS.  I'd argue that this is an acceptable price to pay for a much more intuitive system. In the future, we can mitigate the regression with a fast path that avoids consulting the graph if only one animation is playing. However, in the interest of keeping this patch simple, I didn't do so here.

[the animation composition RFC]: https://github.com/bevyengine/rfcs/blob/main/rfcs/51-animation-composition.md

# Objective

- Describe the objective or issue this PR addresses.
- If you're fixing a specific issue, say "Fixes #X".

## Solution

- Describe the solution used to achieve the objective above.

## Testing

- Did you test these changes? If so, how?
- Are there any parts that need more testing?
- How can other people (reviewers) test your changes? Is there anything specific they need to know?
- If relevant, what platforms did you test these changes on, and are there any important ones you can't test?

---

## Showcase

> This section is optional. If this PR does not include a visual change or does not add a new feature, you can delete this section.

- Help others understand the result of this PR by showcasing your awesome work!
- If this PR adds a new feature or public API, consider adding a brief pseudo-code snippet of it in action
- If this PR includes a visual change, consider adding a screenshot, GIF, or video
  - If you want, you could even include a before/after comparison!
- If the Migration Guide adequately covers the changes, you can delete this section

While a showcase should aim to be brief and digestible, you can use a toggleable section to save space on longer showcases:

<details>
  <summary>Click to view showcase</summary>

```rust
println!("My super cool code.");
```

</details>

## Migration Guide

> This section is optional. If there are no breaking changes, you can delete this section.

- If this PR is a breaking change (relative to the last release of Bevy), describe how a user might need to migrate their code to support these changes
- Simply adding new functionality is not a breaking change.
- Fixing behavior that was definitely a bug, rather than a questionable design choice is not a breaking change.

### CODE REVIEW

File: crates/bevy_animation/src/animation_curves.rs [hunk 1]
```diff
@@ -242,20 +259,72 @@ where
...
     }
 
-    fn apply<'a>(
-        &self,
-        t: f32,
-        _transform: Option<Mut<'a, Transform>>,
-        mut entity: AnimationEntityMut<'a>,
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
```
Comment (aecsocket) [inline crates/bevy_animation/src/animation_curves.rs:264]: I don't think this fn is necessary, since `Reflect` already has `as_any`, and `AnimatableCurve` is `Reflect`.
I tried removing this fn and replacing `AnimatableCurve::as_any` with `Reflect::as_any`, and the animation_graph example seems to work fine, unless I've missed something?

<details>
<summary>Patch</summary>

```diff
diff --git a/crates/bevy_animation/src/animation_curves.rs b/crates/bevy_animation/src/animation_curves.rs
index d8eb8241c..09aa96d14 100644
--- a/crates/bevy_animation/src/animation_curves.rs
+++ b/crates/bevy_animation/src/animation_curves.rs
@@ -78,7 +78,7 @@
 //! [there]: AnimatableProperty

 use core::{
-    any::{Any, TypeId},
+    any::TypeId,
     fmt::{self, Debug, Formatter},
     marker::PhantomData,
 };
@@ -259,10 +259,6 @@ where
         self.curve.domain()
     }

-    fn as_any(&self) -> &dyn Any {
-        self
-    }
-
     fn create_evaluator(&self) -> Box<dyn AnimationCurveEvaluator> {
         Box::new(AnimatableCurveEvaluator {
             evaluator: BasicAnimationCurveEvaluator::default(),
@@ -314,7 +310,7 @@ where
         weight: f32,
         graph_node: AnimationNodeIndex,
     ) -> Result<(), AnimationEvaluationError> {
-        let curve = (*AnimationCurve::as_any(curve))
+        let curve = (*Reflect::as_any(curve))
             .downcast_ref::<AnimatableCurve<P, C>>()
             .unwrap();
         let value = curve.curve.sample_clamped(t);
@@ -357,10 +353,6 @@ where
         Box::new(self.clone())
     }

-    fn as_any(&self) -> &dyn Any {
-        self
-    }
-
     fn domain(&self) -> Interval {
         self.0.domain()
     }
@@ -413,7 +405,7 @@ where
         weight: f32,
         graph_node: AnimationNodeIndex,
     ) -> Result<(), AnimationEvaluationError> {
-        let curve = (*AnimationCurve::as_any(curve))
+        let curve = (*Reflect::as_any(curve))
             .downcast_ref::<TranslationCurve<C>>()
             .unwrap();
         let value = curve.0.sample_clamped(t);
@@ -456,10 +448,6 @@ where
         Box::new(self.clone())
     }

-    fn as_any(&self) -> &dyn Any {
-        self
-    }
-
     fn domain(&self) -> Interval {
         self.0.domain()
     }
@@ -512,7 +500,7 @@ where
         weight: f32,
         graph_node: AnimationNodeIndex,
     ) -> Result<(), AnimationEvaluationError> {
-        let curve = (*AnimationCurve::as_any(curve))
+        let curve = (*Reflect::as_any(curve))
             .downcast_ref::<RotationCurve<C>>()
             .unwrap();
         let value = curve.0.sample_clamped(t);
@@ -555,10 +543,6 @@ where
         Box::new(self.clone())
     }

-    fn as_any(&self) -> &dyn Any {
-        self
-    }
-
     fn domain(&self) -> Interval {
         self.0.domain()
     }
@@ -611,7 +595,7 @@ where
         weight: f32,
         graph_node: AnimationNodeIndex,
     ) -> Result<(), AnimationEvaluationError> {
-        let curve: &ScaleCurve<C> = (*AnimationCurve::as_any(curve)).downcast_ref().unwrap();
+        let curve: &ScaleCurve<C> = (*Reflect::as_any(curve)).downcast_ref().unwrap();
         let value = curve.0.sample_clamped(t);
         self.evaluator
             .stack
@@ -655,10 +639,6 @@ where
         Box::new(self.clone())
     }

-    fn as_any(&self) -> &dyn Any {
-        self
-    }
-
     fn domain(&self) -> Interval {
         self.0.domain()
     }
@@ -711,7 +691,7 @@ where
         weight: f32,
         graph_node: AnimationNodeIndex,
     ) -> Result<(), AnimationEvaluationError> {
-        let curve: &TransformCurve<C> = (*AnimationCurve::as_any(curve)).downcast_ref().unwrap();
+        let curve: &TransformCurve<C> = (*Reflect::as_any(curve)).downcast_ref().unwrap();
         let value = curve.0.sample_clamped(t);
         self.evaluator
             .stack
@@ -784,10 +764,6 @@ where
         self.0.domain()
     }

-    fn as_any(&self) -> &dyn Any {
-        self
-    }
-
     fn create_evaluator(&self) -> Box<dyn AnimationCurveEvaluator> {
         Box::new(WeightsCurveEvaluator {
             stack_morph_target_weights: vec![],
@@ -889,7 +865,7 @@ where
         weight: f32,
         graph_node: AnimationNodeIndex,
     ) -> Result<(), AnimationEvaluationError> {
-        let curve: &WeightsCurve<C> = (*AnimationCurve::as_any(curve)).downcast_ref().unwrap();
+        let curve: &WeightsCurve<C> = (*Reflect::as_any(curve)).downcast_ref().unwrap();

         let prev_morph_target_weights_len = self.stack_morph_target_weights.len();
         self.stack_morph_target_weights
@@ -1003,9 +979,6 @@ pub trait AnimationCurve: Reflect + Debug + Send + Sync {
     /// Returns a boxed clone of this value.
     fn clone_value(&self) -> Box<dyn AnimationCurve>;

-    /// Returns this value upcast to [`Any`].
-    fn as_any(&self) -> &dyn Any;
-
     /// The range of times for which this animation is defined.
     fn domain(&self) -> Interval;
```
</details>

File: crates/bevy_animation/src/animation_curves.rs [hunk 2]
```diff
@@ -429,63 +784,319 @@ where
...
+            .ok_or_else(|| {
+                AnimationEvaluationError::ComponentNotPresent(TypeId::of::<MorphWeights>())
+            })?
+            .weights_mut()
+            .iter_mut()
+            .zip(
+                self.stack_morph_target_weights[(self.stack_morph_target_weights.len()
+                    - self.morph_target_count.unwrap() as usize)..]
+                    .iter(),
+            )
```
Comment (aecsocket) [inline crates/bevy_animation/src/animation_curves.rs:876]: It's somewhat hard to follow what's going on here. Could the index be split into its own variable?

Comment (pcwalton) [reply crates/bevy_animation/src/animation_curves.rs]: Done.

File: crates/bevy_animation/src/animation_curves.rs (hunk 1 above)
Comment (pcwalton) [reply crates/bevy_animation/src/animation_curves.rs]: I removed it.

**Review by mweatherley — APPROVED** [review]
With the concerns about evaluator mapping addressed, this is excellent. The animation graph threading organizes `animate_targets` and `advance_animations` much better, as well.

File: crates/bevy_animation/src/lib.rs [hunk 3]
```diff
@@ -968,70 +929,117 @@ pub fn animate_targets(
...
                 else {
                     continue;
                 };
 
-                let Some(curves) = clip.curves_for_target(target_id) else {
-                    continue;
-                };
+                match animation_graph_node.clip {
+                    None => {
+                        // This is a blend node.
```
[the remark this answers is not in the document]
Comment (pcwalton) [reply crates/bevy_animation/src/lib.rs:957]: This isn't actually true because if all the weights we've seen thus far are zero then the `AnimationCurveEvaluator` won't have been created yet. However, we correctly handle this case by doing nothing.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Great docs and well motivated. Once you and @mweatherley decide what to do with `TransformCurve` (see #15598) and resolve the merge conflicts, ping me and I'll merge this in ASAP.

**Comment by alice-i-cecile** [comment]
>  Yeah you can just delete the entirety of TransformCurve and TransformCurveEvaluator afaik

File: crates/bevy_animation/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mweatherley) [reply crates/bevy_animation/src/lib.rs:957]: Cool! That's good to know.
